### DISCUSSION TITLE
Reuse canvas from node-canvas in Fabric

### DISCUSSION THREAD

**Discussion by konser80** [fabricjs/fabric.js#9497] [Q&A] [answered]
Hi, guys. I'm using node-canvas lib in my project (backend, nodejs server).
And I need some features from Fabric (it can render Text with letter spacing).

So I want to...
- create canvas in node-canvas, do some work there, and
- use that canvas in Fabric, fill text, render it and
- return back to node-canvas to continue using it

Is it possible?
Or any other ways?

**Comment by ShaMan123** [comment, marked as answer]
Create a fabric subclass and do what you need.
You can access the node canvas by canvas.getNodeCanvas() in v6

**Reply by ShaMan123** [reply]
What do you need apart from the canvas API? The API is what node-canvas exposes and is what fabric abstracts over.
To use it you subclass methods such as render, drawObject, _render depending on your needs, _render should fit most needs

**Reply by konser80** [reply]
I use it on a server (nodejs), use only StaticCanvas. I need at least a PDF export.

**Reply by ShaMan123** [reply]
#9185 is a possible solution

**Reply by konser80** [reply]
I'll try it, but codesandbox link is broken =( There is no example

**Reply by ShaMan123** [reply]
Weird
I need to look into it
Clone the repo
Checkout that branch
Run npm start node
And run the app

**Reply by konser80** [reply]
I just realized - no need for a PDF export. Because actually you render an image, so your PDF will contain just an image, not a real objects... =(((

**Reply by ShaMan123** [reply]
Using node canvas PDF exporting you get a full PDF I think, also text and all

**Reply by ShaMan123** [reply]
take a look https://codesandbox.io/p/sandbox/fabric-node-sandbox-forked-qtppjl?file=%2Fsrc%2Findex.mjs%3A27%2C5-34%2C7

**Reply by ShaMan123** [reply]
> But I still can't export pdf

```
const buf = canvas.getNodeCanvas().toBuffer('application/pdf');
fs.writeFileSync('export.pdf', buf);
```

**Reply by konser80** [reply]
Just upgraded to v6 beta...
Still trying to find new registerFont function

**Reply by ShaMan123** [reply]
you need to do that directly from node canvas, not via fabric

**Reply by ShaMan123** [reply]
https://github.com/Automattic/node-canvas#registerfont

**Reply by konser80** [reply]
What version are you using?

I copied your example and got
```pdf.toBuffer is not a function```

tried to console.log(pdf) and looks like it's empty
```HTMLCanvasElement {}```

**Reply by ShaMan123** [reply]
Ahh yes, jsdom does something in the middle
check out the workaround https://codesandbox.io/p/sandbox/fabric-node-sandbox-forked-8hwny3?file=%2Fsrc%2Findex.mjs%3A9%2C1-22%2C2 that doesn't even need #9185

**Reply by konser80** [reply]
That's also doesn't work
File exported, but file data - is a PNG, not PDF

**Reply by ShaMan123** [reply]
This PDF is the downloaded file so I don't understand what you are seeing/doing
[download.pdf](https://github.com/fabricjs/fabric.js/files/13403054/download.pdf)
